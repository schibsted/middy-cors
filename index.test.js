const middy = require('@middy/core');
const createError = require('http-errors');
const middleware = require('./index');

test('Does nothing if has no config', async () => {
    const handler = middy(async () => ({
        statusCode: 200,
        body: JSON.stringify({ foo: 'bar' }),
        headers: {
            someHeader: 'someValue',
        },
    }));

    handler.use(middleware());

    const response = await handler({}, {});
    expect(response).toEqual({
        statusCode: 200,
        headers: {
            someHeader: 'someValue',
        },
        body: JSON.stringify({ foo: 'bar' }),
    });
});

test('Adds CORS headers on success from disallowed origin', async () => {
    const handler = middy(async () => ({
        statusCode: 200,
        body: JSON.stringify({ foo: 'bar' }),
        headers: {
            someHeader: 'someValue',
        },
    }));

    handler.use(
        middleware({
            allowedOrigins: ['https://www.vg.no', 'https://www.tek.no'],
            exposeHeaders: ['x-my-header'],
            maxAge: 2628000, // 1 month
            credentials: true,
            allowMethods: ['GET', 'POST'],
            allowHeaders: ['Content-Type', 'Accept', 'X-Forwarded-For'],
        })
    );

    const response = await handler(
        {
            headers: {
                origin: 'https://www.google.com',
            },
        },
        {}
    );
    expect(response).toEqual({
        statusCode: 200,
        headers: {
            'access-control-allow-credentials': true,
            'access-control-allow-headers': 'Content-Type, Accept, X-Forwarded-For',
            'access-control-allow-methods': 'GET, POST',
            'access-control-expose-headers': 'x-my-header',
            'access-control-max-age': 2628000,
            someHeader: 'someValue',
        },
        body: JSON.stringify({ foo: 'bar' }),
    });
});

test('Adds CORS headers on success from allowed origin', async () => {
    const handler = middy(async () => ({
        statusCode: 200,
        body: JSON.stringify({ foo: 'bar' }),
        headers: {
            someHeader: 'someValue',
        },
    }));

    handler.use(
        middleware({
            allowedOrigins: ['https://www.vg.no', 'https://www.tek.no'],
            exposeHeaders: ['x-my-header'],
            maxAge: 2628000, // 1 month
            credentials: true,
            allowMethods: ['GET', 'POST'],
            allowHeaders: ['Content-Type', 'Accept', 'X-Forwarded-For'],
        })
    );

    const response = await handler(
        {
            headers: {
                origin: 'https://www.tek.no',
            },
        },
        {}
    );
    expect(response).toEqual({
        statusCode: 200,
        headers: {
            'access-control-allow-credentials': true,
            'access-control-allow-headers': 'Content-Type, Accept, X-Forwarded-For',
            'access-control-allow-methods': 'GET, POST',
            'access-control-allow-origin': 'https://www.tek.no',
            'access-control-expose-headers': 'x-my-header',
            'access-control-max-age': 2628000,
            someHeader: 'someValue',
        },
        body: JSON.stringify({ foo: 'bar' }),
    });
});

test('Adds CORS headers on error from disallowed origin', async () => {
    const handler = middy(async () => {
        throw new createError.InternalServerError('whoops');
    });

    handler.use(
        middleware({
            allowedOrigins: ['https://www.vg.no', 'https://www.tek.no'],
            exposeHeaders: ['x-my-header'],
            maxAge: 2628000, // 1 month
            credentials: true,
            allowMethods: ['GET', 'POST'],
            allowHeaders: ['Content-Type', 'Accept', 'X-Forwarded-For'],
        })
    );

    await expect(
        handler(
            {
                headers: {
                    origin: 'https://www.google.com',
                },
            },
            {}
        )
    ).rejects.toEqual(
        expect.objectContaining({
            response: {
                headers: {
                    'access-control-allow-credentials': true,
                    'access-control-allow-headers': 'Content-Type, Accept, X-Forwarded-For',
                    'access-control-allow-methods': 'GET, POST',
                    'access-control-expose-headers': 'x-my-header',
                    'access-control-max-age': 2628000,
                },
            },
        })
    );
});

test('Adds CORS headers on error from allowed origin', async () => {
    const handler = middy(async () => {
        throw new createError.InternalServerError('whoops');
    });

    handler.use(
        middleware({
            allowedOrigins: ['https://www.vg.no', 'https://www.tek.no'],
            exposeHeaders: ['x-my-header'],
            maxAge: 2628000, // 1 month
            credentials: true,
            allowMethods: ['GET', 'POST'],
            allowHeaders: ['Content-Type', 'Accept', 'X-Forwarded-For'],
        })
    );

    await expect(
        handler(
            {
                headers: {
                    origin: 'https://www.tek.no',
                },
            },
            {}
        )
    ).rejects.toEqual(
        expect.objectContaining({
            response: {
                headers: {
                    'access-control-allow-credentials': true,
                    'access-control-allow-headers': 'Content-Type, Accept, X-Forwarded-For',
                    'access-control-allow-methods': 'GET, POST',
                    'access-control-allow-origin': 'https://www.tek.no',
                    'access-control-expose-headers': 'x-my-header',
                    'access-control-max-age': 2628000,
                },
            },
        })
    );
});

test('Keep headers already present in the response on error from disallowed origin', async () => {
    const handler = middy(async () => {
        throw new createError.InternalServerError('whoops');
    });

    // eslint-disable-next-line no-shadow
    handler.onError(async (handler) => {
        // eslint-disable-next-line no-param-reassign
        handler.response = {
            headers: {
                someHeader: 'someValue',
            },
        };

        return true;
    });

    handler.use(
        middleware({
            allowedOrigins: ['https://www.vg.no', 'https://www.tek.no'],
            exposeHeaders: ['x-my-header'],
            maxAge: 2628000, // 1 month
            credentials: true,
            allowMethods: ['GET', 'POST'],
            allowHeaders: ['Content-Type', 'Accept', 'X-Forwarded-For'],
        })
    );

    await expect(
        handler(
            {
                headers: {
                    origin: 'https://www.tek.no',
                },
            },
            {}
        )
    ).rejects.toEqual(
        expect.objectContaining({
            response: {
                headers: {
                    'access-control-allow-credentials': true,
                    'access-control-allow-headers': 'Content-Type, Accept, X-Forwarded-For',
                    'access-control-allow-methods': 'GET, POST',
                    'access-control-allow-origin': 'https://www.tek.no',
                    'access-control-expose-headers': 'x-my-header',
                    'access-control-max-age': 2628000,
                    someHeader: 'someValue',
                },
            },
        })
    );
});
