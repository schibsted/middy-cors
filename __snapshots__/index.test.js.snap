// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adds CORS headers on error 1`] = `
Object {
  "event": Object {
    "headers": Object {
      "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5",
      "connection": "keep-alive",
      "host": "localhost:3000",
      "upgrade-insecure-requests": "1",
      "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
    },
    "httpMethod": "GET",
    "path": "/foobar",
  },
  "response": Object {
    "headers": Object {
      "access-control-allow-credentials": false,
      "access-control-allow-headers": "Content-Type, Accept, X-Forwarded-For",
      "access-control-allow-methods": "GET, POST",
      "access-control-allow-origin": "https://www.vg.no, https://www.tek.no",
      "access-control-expose-headers": "x-my-header",
      "access-control-max-age": 2628000,
    },
    "statusCode": 500,
  },
}
`;

exports[`Adds CORS headers on success 1`] = `
Object {
  "event": Object {
    "headers": Object {
      "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5",
      "connection": "keep-alive",
      "host": "localhost:3000",
      "upgrade-insecure-requests": "1",
      "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
    },
    "httpMethod": "GET",
    "path": "/foobar",
  },
  "response": Object {
    "headers": Object {
      "access-control-allow-credentials": true,
      "access-control-allow-headers": "Content-Type, Accept, X-Forwarded-For",
      "access-control-allow-methods": "GET, POST",
      "access-control-allow-origin": "https://www.vg.no, https://www.tek.no",
      "access-control-expose-headers": "x-my-header",
      "access-control-max-age": 2628000,
    },
    "statusCode": 200,
  },
}
`;

exports[`Does nothing if has no config 1`] = `
Object {
  "event": Object {
    "headers": Object {
      "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5",
      "connection": "keep-alive",
      "host": "localhost:3000",
      "upgrade-insecure-requests": "1",
      "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
    },
    "httpMethod": "GET",
    "path": "/foobar",
  },
  "response": Object {
    "headers": Object {
      "access-control-allow-credentials": false,
    },
    "statusCode": 200,
  },
}
`;
